# Exception

## try-catch-finally
- basic structure
  ```
    try{
        // risky code
    } catch(ExceptionType e){
        // handle
    } finally {
        // always executed
    }
  ```
- {} is mandatory for all try-catch.
- finally block always get executed
- for every try block, one catch / finally block is mandatory.
  ```
  // won't compile. catch / finally block missing
    try {
        // some code
    }
  ```
  ```
  // compile fine. catch block is present
    try{
        // risky code
    } catch(ExceptionType e){
        // handle
    }
  ```
  ```
  // compile fine. finally block is present
    try{
        // risky code
    } finally {
        // always executed
    }
  ```
  ```
  // compile fine. both catch and finally block are present
    try{
        // risky code
    } catch(ExceptionType e){
        // handle
    } finally {
        // always executed
    }
  ```
- consider the following code snippets. The first one won't compile. But the second one is fine.
  ```
  // won't compile, because int x = 2 never execute.
        try {
            throw new RuntimeException();
            int x = 2;
        } catch (RuntimeException e){
            System.out.println(e);
        }
  ```
  ```
  // compile, because exception is inside a condition
        try {
            if(true) throw new RuntimeException();
            int x = 2;
        } catch (RuntimeException e){
            System.out.println(e);
        }
  ```
- chaining catch block
  ```
        try {
            throw new RuntimeException();
        } catch (IllegalArgumentException e){
            System.out.println(e);
        } catch (RuntimeException e){
            System.out.println(e);
        }
  ```
- the chain should be in hierarchy order. here IllegalArgumentException is a subclass of RuntimeException. so it comes first.
- this code won't compile because RuntimeException is superclass of IllegalArgumentException. So it violates the order.
  ```
        try {
            throw new RuntimeException();
        } catch (RuntimeException e){
            System.out.println(e);
        } catch (IllegalArgumentException e){
            System.out.println(e);
        }
  ```
- for chaining catch block, only one will be executed.
- multi catch block
  ```
      try {
          int num = Integer.parseInt(args[1]);
      } catch (ArrayIndexOutOfBoundsException | NumberFormatException e) {
          System.out.println(e);
      }
  ```
- in multi-catch block there can be n number of exceptions
- types of multi-catch must be disjoint
  ```
      try {
          int num = Integer.parseInt(args[1]);
      } catch (ArrayIndexOutOfBoundsException | RuntimeException e) {
          System.out.println(e);
      }
  ```
  this won't compile. because both exceptions are not disjoint. ArrayIndexOutOfBoundsException is a subclass of RuntimeException.


## try-with-resources
- basic structure
  ```
        try (R1 r1; R2 r2; R3 r3 ...) {
            // some code
        } catch (ExceptionType e) {
            // handle
        }
  ```
  ```
        try (FileOutputStream fos = new FileOutputStream("filepath.txt")) {
            String text = "Hello World. This is my java program";
            byte arr[] = text.getBytes();
            fos.write(arr);
        } catch (java.lang.Exception e) {
            System.out.println(e);
        }
  ```
- try-with-resources statement is a try statement that declares one or more resources in it.
- resource is as an object that must be closed after finishing the program
- you can pass any object as a resource that implements **java.lang.AutoCloseable**
- the resource fos will close automatically. so, finally block is optional


## Java Exception Hierarchy
- Exception
  - Checked Exception
    - CloneNotSupportedException
    - InterruptedException
    - IOException
      - FileNotFoundException
      - FileSystemException
        - NoSuchFileException
        - AccessDeniedException
        - NotDirectoryException
        - FileAlreadyExistsException
      - UnknownHostException
      - SocketException
        - ConnectException
    - ReflectiveOperationException
      - ClassNotFoundException
      - IllegalAccessException
      - InstantiationException
      - NoSuchFieldException
      - NoSuchMethodException
  - Unchecked Exception
    - RuntimeException
      - NullPointerException
      - ArithmeticException
      - IllegalArgumentException
        - NumberFormatException
      - IndexOutOfBoundsException
        - ArrayIndexOutOfBoundsException
        - StringIndexOutOfBoundsException
      - NegativeArraySizeException
      - ArrayStoreException
      - ClassCastException
      - IllegalStateException
      - EnumConstantNotPresentException
      - UnsupportedOperationException
      - TypeNotPresentException
      - ConcurrentModificationException


## Java Common Errors
- ErrorInInitializerError
- StackOverflowError
- NoClassDefFoundError
- OutOfMemoryError
